riklet/src/network/net.rs

```diff
@@ -0,0 +1,185 @@
+use std::net::{IpAddr, Ipv4Addr};
+
+use crate::network::tap;
+use devices::virtio::Net as VirtioNet;
+use futures_util::TryStreamExt;
+use rtnetlink::new_connection;
+
+use tracing::debug;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NetworkInterfaceConfig {
+    /// Name of the interface created
+    pub iface_name: String,
+    /// Unique identifier for the network interface, used to compare
+    pub id: String,
+
+    pub ipv4_addr: Ipv4Addr,
+}
+
+impl NetworkInterfaceConfig {
+    pub fn new(id: String, iface_name: String, ipv4_addr: Ipv4Addr) -> Self {
+        NetworkInterfaceConfig {
+            iface_name,
+            id,
+            ipv4_addr,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NetworkInterfaceError {
+    // [devices::virtio::net::Error] already uses thiserror, it's not needed to overwrap our error
+    #[error("{0}")]
+    CreateNetworkDevice(#[from] devices::virtio::net::Error),
+    #[error("Could not open connection to netlink: {0}")]
+    IpSocket(#[from] std::io::Error),
+    #[error("Could not allocate IP address to interface: {0}")]
+    IpAllocation(#[from] rtnetlink::Error),
+}
+
+pub enum NetworkInterface {
+    TapInterface(VirtioNet),
+}
+
+/// An instance of a network implementation, it can be either a tap interface or a just a veth
+/// pair.
+/// # Example
+/// ```
+/// use crate::network::Net;
+/// use crate::network::NetworkInterfaceConfig;
+/// use crate::network::NetworkInterfaceError;
+///
+/// let config = NetworkInterfaceConfig {
+///     iface_name: "test".to_string(),
+///     id: "test".to_string(),
+///    ipv4_addr: Ipv4Addr::new(127, 0, 0, 1),
+/// };
+/// let net = Net::new_with_tap(config).unwrap();
+/// ```
+pub struct Net {
+    /// Unique identifier for the network interface, used to compare
+    pub id: String,
+    /// Type of interface that has been added
+    interface: NetworkInterface,
+}
+
+impl Net {
+    /// Creates a new network interface with a tap interface, it will allocate an IP address to the
+    /// interface depending on the input configuration
+    pub async fn new_with_tap(
+        config: NetworkInterfaceConfig,
+    ) -> Result<Self, NetworkInterfaceError> {
+        debug!("New net tap interface with name: {}", config.iface_name);
```

Comment:
```suggestion
        event!(Level::DEBUG, "New net tap interface with name: {}", config.iface_name);
```

Reply:
`debug!()` is coming from `tokio_tracing`, it's not required to use `event!()`